pkg/build/builder/common.go

```diff
@@ -82,3 +89,107 @@ func updateBuildRevision(c client.BuildInterface, build *api.Build, sourceInfo *
 		glog.Warningf("An error occurred saving build revision: %v", err)
 	}
 }
+
+// randomBuildTag generates a random tag used for building images in such a way
+// that the built image can be referred to unambiguously even in the face of
+// concurrent builds of the same originalTag.
+func randomBuildTag(originalTag string) string {
+	repo, _ := parsers.ParseRepositoryTag(originalTag)
+	suffix := fmt.Sprintf("%08x", rand.Uint32())
+	maxRepoLen := reference.NameTotalLengthMax - len(suffix) - 1
+	if len(repo) > maxRepoLen {
+		repo = repo[:maxRepoLen]
+	}
+	return fmt.Sprintf("%s:%s", repo, suffix)
+}
+
+// runPostCommitHook uses the dockerClient to run the post commit hook of build
+// in a new container based on image. The strategyName is used to generate an
+// appropriate name for the container.
+func runPostCommitHook(dockerClient DockerClient, build *api.Build, strategyName, image string) error {
+	glog.Infof("Running post commit hook with image %s ...", image)
+	containerName := fmt.Sprintf("openshift_%s-build_%s_%s_post-commit_%08x", strategyName, build.Name, build.Namespace, rand.Uint32())
+	return runScriptInContainer(dockerClient, build.Spec.PostCommit.RunBash, image, containerName)
+}
+
+// runScriptInContainer runs a Bash script in an ephemeral Docker container
+// started off image, and returns an error if the script cannot be run or
+// returns a non-zero exit code.
+func runScriptInContainer(client DockerClient, script, image, containerName string) error {
+	limits, err := cgroupResourceLimits()
+	if err != nil {
+		return err
+	}
+	return dockerRun(client, docker.CreateContainerOptions{
+		Name: containerName,
+		Config: &docker.Config{
+			Image:      image,
+			Entrypoint: []string{"bash", "-c"},
+			Cmd:        []string{script},
+		},
+		HostConfig: &docker.HostConfig{
+			// Limit container's resource allocation.
+			CPUShares: limits.CPUShares,
+			CPUPeriod: limits.CPUPeriod,
+			CPUQuota:  limits.CPUQuota,
+			Memory:    limits.Memory,
+		},
+	}, docker.LogsOptions{
+		// Stream logs to stdout and stderr.
+		OutputStream: os.Stdout,
+		ErrorStream:  os.Stderr,
+		Follow:       true,
+		Stdout:       true,
+		Stderr:       true,
+	})
+}
+
+// cgroupResourceLimit represents limits used to constrain resources allocated
+// by containers.
+type cgroupResourceLimit struct {
+	CPUShares int64
+	CPUPeriod int64
+	CPUQuota  int64
+	Memory    int64
+}
+
+// cgroupResourceLimits reads CPU and memory limits from /sys/fs/cgroup.
+func cgroupResourceLimits() (cgroupResourceLimit, error) {
+	r := intFromFileReader{}
+	cpuShares := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.shares")
+	cpuPeriod := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_period_us")
+	cpuQuota := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us")
+	memory := r.read("/sys/fs/cgroup/memory/memory.limit_in_bytes")
+	return cgroupResourceLimit{
+		CPUShares: cpuShares,
+		CPUPeriod: cpuPeriod,
+		CPUQuota:  cpuQuota,
+		Memory:    memory,
+	}, r.err
```

Comment:
I believe I've clarified over IRC that this comparison is oranges to apples.

```
Jan 29 11:45:57 <mfojtik>	rhcarvalho: my point of returning partial struct and error is that you and Ben were discussing returning MAX_INT64, err in case of error
Jan 29 11:46:07 <mfojtik>	rhcarvalho: and I don't see your code different from that
Jan 29 11:46:27 <mfojtik>	rhcarvalho: iow. if there is an error, return empty struct
Jan 29 11:46:33 <mfojtik>	not partially filled struct
Jan 29 12:05:07 <rhcarvalho>	mfojtik: in Ben's code, we're swallowing an error (ErrRange), and the discussion is to what value to return as (V, nil) pair
Jan 29 12:05:37 <rhcarvalho>	mfojtik: we're ignoring the error, thus the question is return X, nil, what is X?
Jan 29 12:05:58 <rhcarvalho>	mfojtik: in the other case that you're complaining, the returned error is non-nil
Jan 29 12:06:12 <rhcarvalho>	mfojtik: therefore we're comparing oranges and apples :/
Jan 29 12:06:47 <mfojtik>	rhcarvalho: ok i misunderstood the Ben case then, I thought we are returning value AND error
```